lightning/src/ln/channelmanager.rs

```diff
@@ -7909,16 +8003,103 @@ where
+								emit_channel_pending_event!(pending_events, chan);
+							}
+						}
+					},
+					_ => try_chan_phase_entry!(self, Err(ChannelError::Close(
+						(
+							"Got an unexpected tx_signatures message".into(),
+							ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
+						))), chan_phase_entry)
+				}
+				Ok(())
+			},
+			hash_map::Entry::Vacant(_) => {
+				Err(MsgHandleErrInternal::send_err_msg_no_close(format!("Got a message for a channel from the wrong node! No such channel for the passed counterparty_node_id {}", counterparty_node_id), msg.channel_id))
+			}
+		}
 	}
 
-	fn internal_tx_abort(&self, counterparty_node_id: &PublicKey, msg: &msgs::TxAbort) {
-		let _: Result<(), _> = handle_error!(self, Err(MsgHandleErrInternal::send_err_msg_no_close(
-			"Dual-funded channels not supported".to_owned(),
-			 msg.channel_id)), *counterparty_node_id);
+	fn maybe_reset_interactive_tx_state(
+		tx_constructor: &mut Option<InteractiveTxConstructor>,
+		channel_id: ChannelId,
+	) -> Option<msgs::TxAbort> {
+		// If we have not reset the negtotiation state yet, we must echo back a `tx_abort`.
+		tx_constructor.take().map(|_| msgs::TxAbort {
+			channel_id,
+			data: "Acknowledged tx_abort".to_string().into_bytes(),
+		})
+	}
+
+	fn internal_tx_abort(&self, counterparty_node_id: &PublicKey, msg: &msgs::TxAbort)
+	-> Result<(), MsgHandleErrInternal> {
+		let per_peer_state = self.per_peer_state.read().unwrap();
+		let peer_state_mutex = per_peer_state.get(counterparty_node_id)
+			.ok_or_else(|| {
+				debug_assert!(false);
+				MsgHandleErrInternal::send_err_msg_no_close(
+					format!("Can't find a peer matching the passed counterparty node_id {}", counterparty_node_id),
+					msg.channel_id)
+			})?;
+		let mut peer_state_lock = peer_state_mutex.lock().unwrap();
+		let peer_state = &mut *peer_state_lock;
+		match peer_state.channel_by_id.entry(msg.channel_id) {
+			hash_map::Entry::Occupied(mut chan_phase_entry) => {
+				let channel_phase = chan_phase_entry.get_mut();
+				let tx_abort_opt = match channel_phase {
+					ChannelPhase::UnfundedInboundV2(chan) => Self::maybe_reset_interactive_tx_state(
+						chan.interactive_tx_constructor_mut(), msg.channel_id),
+					ChannelPhase::UnfundedOutboundV2(chan) => Self::maybe_reset_interactive_tx_state(
+						chan.interactive_tx_constructor_mut(), msg.channel_id),
+					ChannelPhase::Funded(chan) => Self::maybe_reset_interactive_tx_state(
+						&mut chan.interactive_tx_constructor, msg.channel_id),
+					_ => try_chan_phase_entry!(self, Err(ChannelError::Close(
+						(
+							"Got an unexpected tx_signatures message".into(),
+							ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
+						))), chan_phase_entry)
+				};
+				if let Some(tx_abort) = tx_abort_opt {
+					peer_state.pending_msg_events.push(events::MessageSendEvent::SendTxAbort {
+						node_id: *counterparty_node_id,
+						msg: tx_abort,
+					});
+				}
```

Comment:
This would be more readable / cleaner if we drop the helper, IMO, since it has a side effect.

```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index ac6b60c21..100141fa0 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -8052,17 +8052,6 @@ where
                }
        }
 
-       fn maybe_reset_interactive_tx_state(
-               tx_constructor: &mut Option<InteractiveTxConstructor>,
-               channel_id: ChannelId,
-       ) -> Option<msgs::TxAbort> {
-               // If we have not reset the negtotiation state yet, we must echo back a `tx_abort`.
-               tx_constructor.take().map(|_| msgs::TxAbort {
-                       channel_id,
-                       data: "Acknowledged tx_abort".to_string().into_bytes(),
-               })
-       }
-
        fn internal_tx_abort(&self, counterparty_node_id: &PublicKey, msg: &msgs::TxAbort)
        -> Result<(), MsgHandleErrInternal> {
                let per_peer_state = self.per_peer_state.read().unwrap();
@@ -8078,23 +8067,24 @@ where
                match peer_state.channel_by_id.entry(msg.channel_id) {
                        hash_map::Entry::Occupied(mut chan_phase_entry) => {
                                let channel_phase = chan_phase_entry.get_mut();
-                               let tx_abort_opt = match channel_phase {
-                                       ChannelPhase::UnfundedInboundV2(chan) => Self::maybe_reset_interactive_tx_state(
-                                               chan.interactive_tx_constructor_mut(), msg.channel_id),
-                                       ChannelPhase::UnfundedOutboundV2(chan) => Self::maybe_reset_interactive_tx_state(
-                                               chan.interactive_tx_constructor_mut(), msg.channel_id),
-                                       ChannelPhase::Funded(chan) => Self::maybe_reset_interactive_tx_state(
-                                               &mut chan.interactive_tx_constructor, msg.channel_id),
+                               let tx_constructor = match channel_phase {
+                                       ChannelPhase::UnfundedInboundV2(chan) => chan.interactive_tx_constructor_mut(),
+                                       ChannelPhase::UnfundedOutboundV2(chan) => chan.interactive_tx_constructor_mut(),
+                                       ChannelPhase::Funded(chan) => &mut chan.interactive_tx_constructor,
                                        _ => try_chan_phase_entry!(self, Err(ChannelError::Close(
                                                (
                                                        "Got an unexpected tx_signatures message".into(),
                                                        ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
                                                ))), chan_phase_entry)
                                };
-                               if let Some(tx_abort) = tx_abort_opt {
+                               if tx_constructor.take().is_some() {
+                                       let msg = msgs::TxAbort {
+                                               channel_id: msg.channel_id,
+                                               data: "Acknowledged tx_abort".to_string().into_bytes(),
+                                       };
                                        peer_state.pending_msg_events.push(events::MessageSendEvent::SendTxAbort {
                                                node_id: *counterparty_node_id,
-                                               msg: tx_abort,
+                                               msg,
                                        });
                                }
                                Ok(())
```